File: common/split_buffer.cc
// Copyright 2023-2026 David Allison
// All Rights Reserved
// See LICENSE file for licensing information.

#include "common/split_buffer.h"

#include "absl/strings/str_format.h"
#include "common/syscall_shim.h"
#include "common/system_info.h"

#include <cerrno>
#include <cstdint>
#include <cstring>
#include <fcntl.h>
#include <string>
#include <sys/stat.h>
#include <unistd.h>

namespace subspace {
namespace {

absl::Status WriteAll(int fd, const std::string &contents,
                      const std::string &path) {
  const char *p = contents.data();
  size_t remaining = contents.size();
  while (remaining > 0) {
    ssize_t n = write(fd, p, remaining);
    if (n < 0) {
      if (errno == EINTR) {
        continue;
      }
      return absl::InternalError(absl::StrFormat(
          "Failed to write split buffer metadata %s: %s", path,
          strerror(errno)));
    }
    remaining -= static_cast<size_t>(n);
    p += n;
  }
  return absl::OkStatus();
}

uint64_t StableHash64(const std::string &value) {
  uint64_t hash = 1469598103934665603ULL;
  for (unsigned char c : value) {
    hash ^= c;
    hash *= 1099511628211ULL;
  }
  return hash;
}

} // namespace

SplitBufferMetadata SplitBufferMetadataFromProto(
    const ClientBufferHandleMetadataProto &proto) {
  SplitBufferMetadata metadata;
  metadata.channel_name = proto.channel_name();
  metadata.session_id = proto.session_id();
  metadata.buffer_index = proto.buffer_index();
  metadata.slot_id = proto.slot_id();
  metadata.is_prefix = proto.is_prefix();
  metadata.full_size = proto.full_size();
  metadata.allocation_size = proto.allocation_size();
  metadata.handle = static_cast<uintptr_t>(proto.handle());
  metadata.shadow_file = proto.shadow_file();
  metadata.object_name = proto.object_name();
  return metadata;
}

void SplitBufferMetadataToProto(const SplitBufferMetadata &metadata,
                                ClientBufferHandleMetadataProto *proto) {
  proto->set_channel_name(metadata.channel_name);
  proto->set_session_id(metadata.session_id);
  proto->set_buffer_index(metadata.buffer_index);
  proto->set_slot_id(metadata.slot_id);
  proto->set_is_prefix(metadata.is_prefix);
  proto->set_full_size(metadata.full_size);
  proto->set_allocation_size(metadata.allocation_size);
  proto->set_handle(static_cast<uint64_t>(metadata.handle));
  proto->set_shadow_file(metadata.shadow_file);
  proto->set_object_name(metadata.object_name);
}

absl::Status WriteSplitBufferMetadataFile(
    const SplitBufferMetadata &metadata) {
  ClientBufferHandleMetadataProto proto;
  SplitBufferMetadataToProto(metadata, &proto);

  std::string contents;
  if (!proto.SerializeToString(&contents)) {
    return absl::InternalError("Failed to serialize split buffer metadata");
  }

  std::string tmp = metadata.shadow_file + ".tmp";
  int fd = open(tmp.c_str(), O_WRONLY | O_CREAT | O_TRUNC, 0666);
  if (fd == -1) {
    return absl::InternalError(absl::StrFormat(
        "Failed to create split buffer metadata %s: %s", tmp,
        strerror(errno)));
  }

  absl::Status status = WriteAll(fd, contents, tmp);
  if (close(fd) == -1 && status.ok()) {
    status = absl::InternalError(absl::StrFormat(
        "Failed to close split buffer metadata %s: %s", tmp,
        strerror(errno)));
  }
  if (!status.ok()) {
    unlink(tmp.c_str());
    return status;
  }

  if (rename(tmp.c_str(), metadata.shadow_file.c_str()) == -1) {
    unlink(tmp.c_str());
    return absl::InternalError(absl::StrFormat(
        "Failed to publish split buffer metadata %s: %s",
        metadata.shadow_file, strerror(errno)));
  }
  return absl::OkStatus();
}

absl::StatusOr<SplitBufferMetadata>
ReadSplitBufferMetadataFile(const std::string &shadow_file) {
  int fd = open(shadow_file.c_str(), O_RDONLY);
  if (fd == -1) {
    return absl::InternalError(absl::StrFormat(
        "Failed to open split buffer metadata %s: %s", shadow_file,
        strerror(errno)));
  }

  struct stat sb;
  if (fstat(fd, &sb) == -1) {
    close(fd);
    return absl::InternalError(absl::StrFormat(
        "Failed to stat split buffer metadata %s: %s", shadow_file,
        strerror(errno)));
  }

  std::string contents(static_cast<size_t>(sb.st_size), '\0');
  size_t offset = 0;
  while (offset < contents.size()) {
    ssize_t n = read(fd, contents.data() + offset, contents.size() - offset);
    if (n < 0) {
      if (errno == EINTR) {
        continue;
      }
      close(fd);
      return absl::InternalError(absl::StrFormat(
          "Failed to read split buffer metadata %s: %s", shadow_file,
          strerror(errno)));
    }
    if (n == 0) {
      break;
    }
    offset += static_cast<size_t>(n);
  }
  close(fd);

  ClientBufferHandleMetadataProto proto;
  if (!proto.ParseFromString(contents)) {
    return absl::InternalError(absl::StrFormat(
        "Failed to parse split buffer metadata %s", shadow_file));
  }
  return SplitBufferMetadataFromProto(proto);
}

std::string SplitBufferObjectName(const std::string &shadow_file) {
  return absl::StrFormat("subspace_sb_%016x", StableHash64(shadow_file));
}

absl::StatusOr<toolbelt::FileDescriptor>
CreateSplitSharedMemoryBuffer(const SplitBufferMetadata &metadata) {
  int fd = GetSyscallShim().shm_open_fn(metadata.object_name.c_str(),
                                        O_RDWR | O_CREAT | O_EXCL, 0666);
  if (fd == -1) {
    if (errno == EEXIST) {
      return toolbelt::FileDescriptor();
    }
    return absl::InternalError(absl::StrFormat(
        "Failed to create split buffer object %s: %s", metadata.object_name,
        strerror(errno)));
  }

  toolbelt::FileDescriptor shm_fd(fd);
  uint64_t allocation_size =
      metadata.allocation_size != 0 ? metadata.allocation_size
                                    : PageAlignedSize(metadata.full_size);
  if (GetSyscallShim().ftruncate_fn(shm_fd.Fd(),
                                    static_cast<off_t>(allocation_size)) == -1) {
    (void)GetSyscallShim().shm_unlink_fn(metadata.object_name.c_str());
    return absl::InternalError(absl::StrFormat(
        "Failed to size split buffer object %s: %s", metadata.object_name,
        strerror(errno)));
  }

  return shm_fd;
}

absl::StatusOr<toolbelt::FileDescriptor>
OpenSplitSharedMemoryBuffer(const SplitBufferMetadata &metadata, int flags) {
  int fd = GetSyscallShim().shm_open_fn(metadata.object_name.c_str(), flags,
                                        0666);
  if (fd == -1) {
    return absl::InternalError(absl::StrFormat(
        "Failed to open split buffer object %s: %s", metadata.object_name,
        strerror(errno)));
  }
  return toolbelt::FileDescriptor(fd);
}

absl::Status DestroySplitSharedMemoryBuffer(
    const SplitBufferMetadata &metadata) {
  absl::Status status = absl::OkStatus();
  if (GetSyscallShim().shm_unlink_fn(metadata.object_name.c_str()) == -1 &&
      errno != ENOENT) {
    status = absl::InternalError(absl::StrFormat(
        "Failed to unlink split buffer object %s: %s", metadata.object_name,
        strerror(errno)));
  }
  if (unlink(metadata.shadow_file.c_str()) == -1 && errno != ENOENT &&
      status.ok()) {
    status = absl::InternalError(absl::StrFormat(
        "Failed to remove split buffer metadata %s: %s", metadata.shadow_file,
        strerror(errno)));
  }
  return status;
}

} // namespace subspace

File: rust_client/src/split_buffer.rs
// Copyright 2026 David Allison
// All Rights Reserved
// See LICENSE file for licensing information.

use crate::error::{Result, SubspaceError};
use crate::proto;
use crate::syscall_shim::{shim_close, shim_ftruncate, shim_open, shim_read, shim_write};
use nix::fcntl::OFlag;
use nix::sys::stat::Mode;
use prost::Message as ProstMessage;
use std::os::unix::io::RawFd;
use std::sync::Arc;

#[derive(Debug, Clone, Default)]
pub struct SplitBufferMetadata {
    pub channel_name: String,
    pub session_id: u64,
    pub buffer_index: u32,
    pub slot_id: u32,
    pub is_prefix: bool,
    pub full_size: u64,
    pub allocation_size: u64,
    pub handle: u64,
    pub shadow_file: String,
    pub object_name: String,
}

#[derive(Debug, Clone, Copy)]
pub struct SplitBufferMapping {
    pub handle: u64,
    pub address: *mut u8,
    pub size: usize,
    pub private_data: usize,
}

impl Default for SplitBufferMapping {
    fn default() -> Self {
        Self {
            handle: 0,
            address: std::ptr::null_mut(),
            size: 0,
            private_data: 0,
        }
    }
}

pub type SplitBufferAllocateCallback =
    Arc<dyn Fn(&SplitBufferMetadata) -> Result<SplitBufferMapping> + Send + Sync>;
pub type SplitBufferMapCallback =
    Arc<dyn Fn(&SplitBufferMetadata) -> Result<SplitBufferMapping> + Send + Sync>;
pub type SplitBufferUnmapCallback =
    Arc<dyn Fn(&SplitBufferMetadata, &SplitBufferMapping) -> Result<()> + Send + Sync>;
pub type SplitBufferFreeCallback =
    Arc<dyn Fn(&SplitBufferMetadata, &SplitBufferMapping) -> Result<()> + Send + Sync>;

#[derive(Clone, Default)]
pub struct SplitBufferCallbacks {
    pub allocate: Option<SplitBufferAllocateCallback>,
    pub map: Option<SplitBufferMapCallback>,
    pub unmap: Option<SplitBufferUnmapCallback>,
    pub free: Option<SplitBufferFreeCallback>,
}

impl std::fmt::Debug for SplitBufferCallbacks {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("SplitBufferCallbacks")
            .field("allocate", &self.allocate.is_some())
            .field("map", &self.map.is_some())
            .field("unmap", &self.unmap.is_some())
            .field("free", &self.free.is_some())
            .finish()
    }
}

impl SplitBufferMetadata {
    pub fn to_proto(&self, allocator: &str) -> proto::ClientBufferHandleMetadataProto {
        let allocator = match allocator {
            "split_shm" => proto::ClientBufferAllocator::SplitShm as i32,
            "split_callback" => proto::ClientBufferAllocator::SplitCallback as i32,
            "split_buffer_free_test" => {
                proto::ClientBufferAllocator::SplitBufferFreeTest as i32
            }
            _ => proto::ClientBufferAllocator::Unspecified as i32,
        };
        proto::ClientBufferHandleMetadataProto {
            channel_name: self.channel_name.clone(),
            session_id: self.session_id,
            buffer_index: self.buffer_index,
            slot_id: self.slot_id,
            is_prefix: self.is_prefix,
            full_size: self.full_size,
            allocation_size: self.allocation_size,
            handle: self.handle,
            shadow_file: self.shadow_file.clone(),
            object_name: self.object_name.clone(),
            allocator,
        }
    }
}

impl From<proto::ClientBufferHandleMetadataProto> for SplitBufferMetadata {
    fn from(proto: proto::ClientBufferHandleMetadataProto) -> Self {
        Self {
            channel_name: proto.channel_name,
            session_id: proto.session_id,
            buffer_index: proto.buffer_index,
            slot_id: proto.slot_id,
            is_prefix: proto.is_prefix,
            full_size: proto.full_size,
            allocation_size: proto.allocation_size,
            handle: proto.handle,
            shadow_file: proto.shadow_file,
            object_name: proto.object_name,
        }
    }
}

pub fn page_aligned_size(size: u64) -> u64 {
    let page_size = unsafe { libc::sysconf(libc::_SC_PAGESIZE) };
    let page_size = if page_size <= 0 {
        4096
    } else {
        page_size as u64
    };
    (size + page_size - 1) & !(page_size - 1)
}

pub fn split_buffer_object_name(shadow_file: &str) -> String {
    format!("subspace_sb_{:016x}", stable_hash64(shadow_file))
}

fn stable_hash64(value: &str) -> u64 {
    let mut hash = 1_469_598_103_934_665_603u64;
    for c in value.bytes() {
        hash ^= c as u64;
        hash = hash.wrapping_mul(1_099_511_628_211u64);
    }
    hash
}

pub fn write_split_buffer_metadata_file(metadata: &SplitBufferMetadata) -> Result<()> {
    let proto = metadata.to_proto("");
    let mut contents = Vec::new();
    proto.encode(&mut contents)?;

    let tmp = format!("{}.tmp", metadata.shadow_file);
    let fd = shim_open(
        &tmp,
        OFlag::O_WRONLY | OFlag::O_CREAT | OFlag::O_TRUNC,
        Mode::from_bits_truncate(0o666),
    )?;
    let write_result = write_all(fd, &contents);
    let close_result = shim_close(fd);
    if let Err(e) = write_result {
        let _ = std::fs::remove_file(&tmp);
        return Err(e);
    }
    if close_result == -1 {
        let _ = std::fs::remove_file(&tmp);
        return Err(SubspaceError::Io(std::io::Error::last_os_error()));
    }
    std::fs::rename(&tmp, &metadata.shadow_file)?;
    Ok(())
}

pub fn read_split_buffer_metadata_file(shadow_file: &str) -> Result<SplitBufferMetadata> {
    let mut last_error = None;
    for _ in 0..100 {
        match read_split_buffer_metadata_file_once(shadow_file) {
            Ok(metadata) => return Ok(metadata),
            Err(e) => {
                last_error = Some(e);
                std::thread::sleep(std::time::Duration::from_millis(1));
            }
        }
    }
    Err(last_error.unwrap_or_else(|| {
        SubspaceError::Internal(format!(
            "Failed to read split buffer metadata {shadow_file}"
        ))
    }))
}

fn read_split_buffer_metadata_file_once(shadow_file: &str) -> Result<SplitBufferMetadata> {
    let fd = shim_open(shadow_file, OFlag::O_RDONLY, Mode::empty())?;
    let result = (|| -> Result<SplitBufferMetadata> {
        let stat = crate::syscall_shim::shim_fstat(fd)?;
        let mut contents = vec![0u8; stat.st_size as usize];
        let mut offset = 0usize;
        while offset < contents.len() {
            let n = shim_read(
                fd,
                contents[offset..].as_mut_ptr() as *mut libc::c_void,
                contents.len() - offset,
            );
            if n < 0 {
                return Err(SubspaceError::Io(std::io::Error::last_os_error()));
            }
            if n == 0 {
                break;
            }
            offset += n as usize;
        }
        let proto = proto::ClientBufferHandleMetadataProto::decode(&contents[..offset])?;
        Ok(proto.into())
    })();
    shim_close(fd);
    result
}

fn write_all(fd: RawFd, contents: &[u8]) -> Result<()> {
    let mut offset = 0usize;
    while offset < contents.len() {
        let n = shim_write(
            fd,
            contents[offset..].as_ptr() as *const libc::c_void,
            contents.len() - offset,
        );
        if n < 0 {
            return Err(SubspaceError::Io(std::io::Error::last_os_error()));
        }
        offset += n as usize;
    }
    Ok(())
}

pub fn create_split_shared_memory_buffer(metadata: &SplitBufferMetadata) -> Result<Option<RawFd>> {
    let fd = split_shm_open(
        &metadata.object_name,
        OFlag::O_RDWR | OFlag::O_CREAT | OFlag::O_EXCL,
        Mode::from_bits_truncate(0o666),
    );
    let fd = match fd {
        Ok(fd) => fd,
        Err(nix::errno::Errno::EEXIST) => return Ok(None),
        Err(e) => return Err(e.into()),
    };
    let allocation_size = if metadata.allocation_size == 0 {
        page_aligned_size(metadata.full_size)
    } else {
        metadata.allocation_size
    };
    if shim_ftruncate(fd, allocation_size as libc::off_t) == -1 {
        let _ = shim_close(fd);
        return Err(SubspaceError::Io(std::io::Error::last_os_error()));
    }
    Ok(Some(fd))
}

pub fn open_split_shared_memory_buffer(metadata: &SplitBufferMetadata) -> Result<RawFd> {
    Ok(split_shm_open(
        &metadata.object_name,
        OFlag::O_RDWR,
        Mode::empty(),
    )?)
}

#[cfg(target_os = "linux")]
fn split_shm_open(name: &str, oflag: OFlag, mode: Mode) -> nix::Result<RawFd> {
    shim_open(&format!("/dev/shm/{name}"), oflag, mode)
}

#[cfg(not(target_os = "linux"))]
fn split_shm_open(name: &str, oflag: OFlag, mode: Mode) -> nix::Result<RawFd> {
    crate::syscall_shim::shim_shm_open(name, oflag, mode)
}
